backend/src/halo2/circuit_builder.rs

```diff
@@ -84,25 +85,34 @@ pub(crate) struct PowdrCircuit<'a, T> {
     witgen_callback: Option<WitgenCallback<T>>,
 }
 
+fn get_publics<T: FieldElement>(analyzed: &Analyzed<T>) -> Vec<(String, usize)> {
+    let mut publics = analyzed
+        .public_declarations
+        .values()
+        .map(|public_declaration| {
+            let witness_name = public_declaration.referenced_poly_name();
+            let witness_offset = public_declaration.index as usize;
+            (witness_name, witness_offset)
+        })
+        .collect::<Vec<_>>();
+
+    // Sort, so that the order is deterministic
+    publics.sort();
+    publics
+}
+
 impl<'a, T: FieldElement> PowdrCircuit<'a, T> {
     pub(crate) fn new(analyzed: &'a Analyzed<T>, fixed: &'a [(String, Vec<T>)]) -> Self {
-        let mut publics = analyzed
-            .public_declarations
-            .values()
-            .map(|public_declaration| {
-                let witness_name = public_declaration.referenced_poly_name();
-                let witness_offset = public_declaration.index as usize;
-                (witness_name, witness_offset)
-            })
-            .collect::<Vec<_>>();
-        // Sort, so that the order is deterministic
-        publics.sort();
+        for (fixed_name, _) in fixed {
+            assert!(fixed_name != FIRST_STEP_NAME);
```

Comment:
You could also check whether `first_step` is already there, but also fine like this

Reply:
There is a difference between the one that's often there and the one we're adding here: The `__first_step` is only 1 once, and a column like `first_step = [1] + [0]*` would be 1 twice, in row 0 and row `degree`, because we double the degree in the Halo2 backend implementation!

This way, you can use constraints like `first_step' * (x - 1) = 0` to assert some value in the *last* row, because we simulate the wrapping. But in the `__first_step` column, we actually want it to be active in one row only.